lua/diagnostics.lua

```diff
@@ -0,0 +1,46 @@
+local module = {}
+
+local ale_type_to_diagnostic_severity = {
+  E = vim.diagnostic.severity.ERROR,
+  W = vim.diagnostic.severity.WARN,
+  I = vim.diagnostic.severity.INFO
+}
+
+module.sendAleResultsToDiagnostics = function(buffer, loclist)
+  diagnostics = {}
+
+  -- Convert all the ALE loclist items to the shape that Neovim's diagnostic
+  -- API is expecting.
+  for _, location in ipairs(loclist) do
+    table.insert(
+      diagnostics,
+      -- All line numbers from ALE are 1-indexed, but all line numbers
+      -- in the diagnostics API are 0-indexed, so we have to subtract 1
+      -- to make this work.
+      {
+        lnum = location.lnum - 1,
+        -- Ending line number, or if we don't have one, just make it the same
+        -- as the starting line number
+        end_lnum = (location.end_lnum or location.lnum) - 1,
+        -- Which column does the error start on?
+        col = (location.col or 1) - 1,
```

Comment:
Could it be that this one is not correct. Here 'trouble' has trouble:
<img width="1231" alt="image" src="https://user-images.githubusercontent.com/1944737/166224134-10cc0c46-d60b-46e1-be2f-323a0d09974a.png">

Reply:
column_number is -1 in `trouble/view.lua:453`. In ale `lua/diagnostics.lua` location.col is 0, lua interpreter says that `0 or 1` is 0, so your code sets column as -1.

Reply:
Thanks! I will fix this error. Do you have a line of text that reproduces it? I'm just curious about where the error is supposed to be in your case. Something tells me it starts in column 0?

Reply:
(a screenshot of the broken line of code in your editor would be fine)

Reply:
Here is minimal reproducible Typescript case. There will be error on line 3 because import statement is incomplete. So, you are right that it is column 0 (it is interesting that eslint and tsserver reports different column):

```
import

const x = 1;
console.log(x);
```

<img width="997" alt="image" src="https://user-images.githubusercontent.com/1944737/166317820-58704925-3866-4315-8c1c-f9e5bf3b05b5.png">

Reply:
Last questions:

1. For `tsserver`, is that via Neovim LSP, or ALE's LSP? Just curious about the column discrepancy between tsserver + eslint.
2. For `eslint`, is that just configured with the standard ALE setup? Do you use `eslint_d` or `eslint`?

Reply:
1. ALE's LSP.
2. Standard ALE setup. `eslint`

Reply:
This is clearly some kind of `eslint` corner case as the same code in `js` file produces different indexes so I assume this depends even on how typescript is handled. I guess it would be good not to pass `-1` as column (anyway that does not make sense).

Reply:
Hmm… is the thing to do to wrap it in in `Math.max(0, lineNumber)` to ensure it never drops below `0`? Or is there a safer approach to handle this edge case, I wonder.

Reply:
KISS `col = math.max((location.col or 1) - 1, 0),`

Reply:
I often use "max" and "min" functions in various places with the last line number, the end of a line, and so on. Often tools just report numbers that are wrong. Look out for "off-by-one" issues.